main-service/src/main/java/ru/practicum/mainService/compilation/service/CompilationServiceImpl.java

```diff
@@ -0,0 +1,146 @@
+    @Transactional
+    public CompilationResponseDto update(CompilationRequestDto compilationRequestDto, Long compId) {
+        Compilation compilation = compExistsAndGet(compId);
+
+        if (compilationRequestDto.getPinned() != null) {
+            compilation.setPinned(compilationRequestDto.getPinned());
+        }
+        if (compilationRequestDto.getEvents() != null) {
+            Set<Event> events = eventsListToSet(eventRepository.findEventsByIdIn(compilationRequestDto.getEvents()));
+            allEventsExist(events.size(), compilationRequestDto.getEvents().size());
+            compilation.setEvents(events);
+        }
+        if (compilationRequestDto.getTitle() != null) {
+            compilation.setTitle(compilationRequestDto.getTitle());
+        }
+
+        log.info("Обновлена compilation с ID {} на параметры {}", compId, compilationRequestDto);
+        return compilationMapper.toCompilationDto(compilation, findViewsAndRequestsForEvents(compilation.getEvents()));
+    }
+
+    @Override
+    @Transactional
+    public void delete(Long compId) {
+        compExists(compId);
+        log.info("Compilation с ID {} удалена.", compId);
+        compilationRepository.deleteById(compId);
+    }
+
+    @Override
+    @Transactional(readOnly = true)
+    public CompilationResponseDto getById(Long compId) {
+        Compilation compilation = compExistsAndGet(compId);
+        Set<Event> events = compilation.getEvents();
+        CompilationResponseDto compilationResponseDto = compilationMapper.toCompilationDto(compilation,
+                findViewsAndRequestsForEvents(events));
+        log.info("Получена Compilation с ID {}.", compId);
+        return compilationResponseDto;
+    }
+
+    @Override
+    @Transactional(readOnly = true)
+    public List<CompilationResponseDto> getAll(Boolean pinned, Integer from, Integer size) {
+        List<Compilation> compilations;
+
+        if (pinned != null) {
+            compilations = compilationRepository.findAllByPinnedIs(pinned, PageRequest.of(from / size, size));
+        } else {
+            compilations = compilationRepository.findAll(PageRequest.of(from / size, size)).toList();
+        }
+
+        log.info("Получен список compilations с параметрами pinned={}, from={}, size={}", pinned, from, size);
+        List<CompilationResponseDto> result = new ArrayList<>();
+        for (Compilation compilation : compilations) {
+            result.add(compilationMapper.toCompilationDto(compilation,
+                    findViewsAndRequestsForEvents(compilation.getEvents())));
+        }
+        return result;
+    }
+
+    private List<EventShortDto> findViewsAndRequestsForEvents(Set<Event> events) {
+        Map<Long, Long> views = statService.getViews(new ArrayList<>(events));
+        Map<Long, Long> confirmedRequests = statService.getConfirmedRequests(new ArrayList<>(events));
+
+        List<EventShortDto> result = new ArrayList<>();
+        for (Event event : events) {
```

Comment:
Используй [stream map](https://www.geeksforgeeks.org/stream-map-java-examples/):
```suggestion
         return events.stream().map(
                event ->
                    eventMapper.toEventShortDto(
                        event, confirmedRequests.getOrDefault(event.getId(), 0L),
                        views.getOrDefault(event.getId(), 0L)
                    )
            )
            .collect(Collectors.toList());
```